contracts/REXTwoWayUniswapMarket.sol

```diff
@@ -0,0 +1,614 @@
+        require(shares == 0, "Already streaming");
+        (, , shares, ) = getIDAShares(OUTPUTB_INDEX, shareholder);
+        require(shares == 0, "Already streaming");
+    }
+
+    function beforeAgreementTerminated(
+        ISuperToken _superToken,
+        address _agreementClass,
+        bytes32, //_agreementId,
+        bytes calldata _agreementData,
+        bytes calldata _ctx
+    ) external view virtual override returns (bytes memory _cbdata) {
+        _onlyHost();
+        if (!_isInputToken(_superToken) || !_isCFAv1(_agreementClass))
+            return _ctx;
+
+        (
+            address _shareholder,
+            int96 _flowRateMain,
+            uint256 _timestamp
+        ) = _getShareholderInfo(_agreementData, _superToken);
+
+        uint256 _uinvestAmount = _calcUserUninvested(
+            _timestamp,
+            uint256(uint96(_flowRateMain)),
+            // Select the correct lastDistributionAt for this _superToken
+            _getLastDistributionAt(_superToken)
+        );
+        _cbdata = abi.encode(_uinvestAmount, int256(_flowRateMain));
+    }
+
+    function _getEncodedPath(address[] memory path, uint24[] memory poolFees)
+        internal
+        view
+        returns (bytes memory encodedPath)
+    {
+        for (uint256 i = 0; i < path.length; i++) {
+            if (i == path.length - 1) {
+                encodedPath = abi.encodePacked(encodedPath, path[i]);
+            } else {
+                encodedPath = abi.encodePacked(
+                    encodedPath,
+                    path[i],
+                    poolFees[i]
+                );
+            }
+        }
+        return encodedPath;
+    }
+
+    function _swap(
+        ISuperToken input,
+        ISuperToken output,
+        uint256 amount
+    ) internal returns (uint256) {
+        address inputToken; // The underlying input token address
+        address outputToken; // The underlying output token address
+        uint256 minOutput; // The minimum amount of output tokens based on Tellor
+        uint256 outputAmount; // The balance before the swap
+
+        inputToken = input.getUnderlyingToken();
+        outputToken = output.getUnderlyingToken();
+
+        // Downgrade and scale the input amount
+        // Handle case input or output is native supertoken
+        if (inputToken == address(0)) {
+            inputToken = address(input);
+        } else {
+            input.downgrade(amount);
+            // Scale it to 1e18 for calculations
+            amount =
+                ERC20(inputToken).balanceOf(address(this)) *
+                (10**(18 - ERC20(inputToken).decimals()));
+        }
+
+        if (outputToken == address(0)) {
+            outputToken = address(output);
+        }
+
+        minOutput =
+            (amount * market.oracles[input].usdPrice) /
+            market.oracles[output].usdPrice;
+        minOutput = (minOutput * (1e6 - market.rateTolerance)) / 1e6;
```

Comment:
Also, not sure this is the right way to do this since the _swap method should be calculating using a "just in time" call to `oracle.observe()`